fraud/service_test.go

```diff
@@ -56,18 +60,216 @@ func TestService_Broadcast(t *testing.T) {
+	require.NoError(t, err)
+
+	_, err = subsB.Proof(ctx)
+	require.NoError(t, err)
+
+	newCtx, cancel := context.WithTimeout(ctx, time.Millisecond*500)
+	t.Cleanup(cancel)
+	_, err = subsC.Proof(newCtx)
+	require.Error(t, err)
+	require.True(t, bl.Contains(net.Hosts()[1].ID()))
+}
+
+func TestService_GossipingOfFaultBEFP(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer t.Cleanup(cancel)
+	// create mock network
+	net, err := mocknet.FullMeshLinked(3)
+	require.NoError(t, err)
+	bServ := mdutils.Bserv()
+
+	// create first fraud service that will broadcast incorrect Fraud Proof
+	serviceA, store1 := createServiceWithHost(t, net.Hosts()[0])
+
+	h, err := store1.GetByHeight(context.TODO(), 1)
+	require.NoError(t, err)
+
+	// create and break byzantine error
+	_, err = generateByzantineError(ctx, t, h, bServ)
+	require.Error(t, err)
+	var errByz *ipld.ErrByzantine
+	require.True(t, errors.As(err, &errByz))
+	errByz.Index = 2
+
+	fserviceA := serviceA.(*service)
+	require.NotNil(t, fserviceA)
+
+	bl, err := pubsub.NewTimeCachedBlacklist(time.Hour)
+	require.NoError(t, err)
+	// create pub sub in order to listen for Fraud Proof
+	psB, err := pubsub.NewGossipSub(ctx, net.Hosts()[1], // -> B
+		pubsub.WithMessageSignaturePolicy(pubsub.StrictNoSign), pubsub.WithBlacklist(bl))
+	require.NoError(t, err)
+	// create second service that will receive and validate Fraud Proof
+	serviceB := NewService(psB, store1.GetByHeight, sync.MutexWrap(datastore.NewMapDatastore()))
+	fserviceB := serviceB.(*service)
+	require.NotNil(t, fserviceB)
+	addrB := host.InfoFromHost(net.Hosts()[1]) // -> B
+
+	// create pub sub in order to listen for Fraud Proof
+	psC, err := pubsub.NewGossipSub(ctx, net.Hosts()[2], // -> C
+		pubsub.WithMessageSignaturePolicy(pubsub.StrictNoSign))
+	require.NoError(t, err)
+	serviceC := NewService(psC, store1.GetByHeight, sync.MutexWrap(datastore.NewMapDatastore()))
+
+	// perform subscriptions
+	sub0, err := net.Hosts()[0].EventBus().Subscribe(&event.EvtPeerIdentificationCompleted{})
+	require.NoError(t, err)
+	sub2, err := net.Hosts()[2].EventBus().Subscribe(&event.EvtPeerIdentificationCompleted{})
+	require.NoError(t, err)
+
+	// establish connections
+	// connect peers: A -> B -> C, so A and C are not connected to each other
+	require.NoError(t, net.Hosts()[0].Connect(ctx, *addrB)) // host[0] is A
+	require.NoError(t, net.Hosts()[2].Connect(ctx, *addrB)) // host[2] is C
+
+	// wait on both peer identification events
+	for i := 0; i < 2; i++ {
+		select {
+		case <-sub0.Out():
+		case <-sub2.Out():
+		case <-ctx.Done():
+			assert.FailNow(t, "timeout waiting for peers to connect")
+		}
+	}
+
+	// subscribe to BEFP
+	_, err = serviceA.Subscribe(BadEncoding)
+	require.NoError(t, err)
+
+	subsB, err := serviceB.Subscribe(BadEncoding)
+	require.NoError(t, err)
+	defer subsB.Cancel()
+
+	subsC, err := serviceC.Subscribe(BadEncoding)
+	require.NoError(t, err)
+	defer subsC.Cancel()
```

Comment:
Cancel for A?

Reply:
Same for another test

Reply:
`subsA` is not created: `_, err = serviceA.Subscribe(BadEncoding)`

Reply:
Well, it is created because you call the func, you just ignore the value

Reply:
technically, yes. but should we still cancel it in this case?

Reply:
and technically we should cancel. We should be nice to our code ;)